nixos/modules/services/misc/gitlab.nix

```diff
@@ -626,46 +891,91 @@ in {
-
-            # enable required pg_trgm extension for gitlab
-            ${pkgs.sudo}/bin/sudo -u ${pgSuperUser} psql ${cfg.databaseName} -c "CREATE EXTENSION IF NOT EXISTS pg_trgm"
+        cp -f ${cfg.packages.gitlab}/share/gitlab/VERSION ${cfg.statePath}/VERSION
+        rm -rf ${cfg.statePath}/db/*
+        cp -rf --no-preserve=mode ${cfg.packages.gitlab}/share/gitlab/config.dist/* ${cfg.statePath}/config
+        cp -rf --no-preserve=mode ${cfg.packages.gitlab}/share/gitlab/db/* ${cfg.statePath}/db
+
+        ${cfg.packages.gitlab-shell}/bin/install
+
+        ${optionalString cfg.smtp.enable ''
+          install -m u=rw ${smtpSettings} ${cfg.statePath}/config/initializers/smtp_settings.rb
+          ${optionalString (cfg.smtp.passwordFile != null) ''
+            smtp_password=$(<'${cfg.smtp.passwordFile}')
+            ${pkgs.replace}/bin/replace-literal -e '@smtpPassword@' "$smtp_password" '${cfg.statePath}/config/initializers/smtp_settings.rb'
+          ''}
+        ''}
+
+        (
+          umask u=rwx,g=,o=
+
+          ${pkgs.openssl}/bin/openssl rand -hex 32 > ${cfg.statePath}/gitlab_shell_secret
+
+          ${genSecretsReplacementSnippet
+              (recursiveGetAttrWithJqPrefix gitlabConfig "_secret")
+              (pkgs.writeText "gitlab.yml" (builtins.toJSON gitlabConfig))
+              "${cfg.statePath}/config/gitlab.yml"
+          }
+
+          if [[ -h '${cfg.statePath}/config/secrets.yml' ]]; then
+            rm '${cfg.statePath}/config/secrets.yml'
           fi
+          export secret="${if cfg.secrets.secretFile != null then "$(<'${cfg.secrets.secretFile}')" else cfg.secrets.secret}"
+          export db="${if cfg.secrets.dbFile != null then "$(<'${cfg.secrets.dbFile}')" else cfg.secrets.db}"
+          export otp="${if cfg.secrets.otpFile != null then "$(<'${cfg.secrets.otpFile}')" else cfg.secrets.otp}"
+          export jws="${if cfg.secrets.jwsFile != null then "$(<'${cfg.secrets.jwsFile}')" else cfg.secrets.jws}"
+          ${pkgs.jq}/bin/jq -n '{production: {secret_key_base: $ENV.secret,
+                                              otp_key_base: $ENV.db,
+                                              db_key_base: $ENV.otp,
+                                              openid_connect_signing_key: $ENV.jws}}' \
+                            > '${cfg.statePath}/config/secrets.yml'
+        )
+
+        initial_root_password="${if cfg.initialRootPasswordFile != null then "$(<'${cfg.initialRootPasswordFile}')" else cfg.initialRootPassword}"
+        ${gitlab-rake}/bin/gitlab-rake gitlab:db:configure GITLAB_ROOT_PASSWORD="$initial_root_password" \
+                                                           GITLAB_ROOT_EMAIL='${cfg.initialRootEmail}'
+
+        ${pkgs.git}/bin/git config --global core.autocrlf "input"
+      '';
 
-          ${pkgs.sudo}/bin/sudo -u ${cfg.user} -H ${gitlab-rake}/bin/gitlab-rake db:schema:load
+      preStartFullPrivileges = ''
```

Comment:
It seems to me that almost all of the code in here can (and should) be taken care of in a declarative way.

- Any password files should already have read permissions for the user that requires them.
- Databases should be provisioned with `services.postgresql.ensureDatabases` and database users with `services.postgresql.ensureUsers`.
- I'm not familiar with `pgsql` extensions, but it seems like maybe there should be a `services.postgresql.ensureExtensions` option...?

I'll also mention that I'm a fan of a `services.foobar.database.createLocally` option to indicate that the sysadmin wants a database provisioned for them.

Reply:
While I do agree that it would be preferable to do everything in a declarative way, I'm not entirely sure this is possible. I'll just reply to the points in the same order:
- I'm not sure what this is in reference to?
- The problem is that there's no way, afaict, to create a user with a password in a secure fashion using the existing postgres service options. Maybe the postgres module should provide a way, though?
- Same here, as you say, it's not currently possible, but maybe it could be implemented?

Yeah, I guess we could introduce a `createLocally` option and rely on it instead of checking `databaseHost`.

Reply:
> Any password files should already have read permissions for the user that requires them.

My intention was to point out operations involving the `passwordFile` should not require `root` privileges, generally speaking.

> The problem is that there's no way, afaict, to create a user with a password in a secure fashion using the existing postgres service options. Maybe the postgres module should provide a way, though?

I'm suggesting that passwords are inferior to socket authentication, so socket authentication should be preferred when provisioning a database locally.

> Same here, as you say, it's not currently possible, but maybe it could be implemented?

From the looks of your code it looks like it could be implemented fairly easily. I believe I've seen extensions provisioned this way in several modules, so I think this would be valuable.

> Yeah, I guess we could introduce a createLocally option and rely on it instead of checking databaseHost.

Some people may wish to provision their own database, or be migrating from an existing install so `createLocally` is nice to have.

Reply:
> My intention was to point out operations involving the `passwordFile` should not require `root` privileges, generally speaking.

Ah, I see.

> I'm suggesting that passwords are inferior to socket authentication, so socket authentication should be preferred when provisioning a database locally.

True, but socket authentication isn't always possible. For gitlab it seems like it is, though, so I should experiment a bit and see if a switch to it is doable.

> From the looks of your code it looks like it could be implemented fairly easily. I believe I've seen extensions provisioned this way in several modules, so I think this would be valuable.

Yes, probably. I'll have to look at the postgres module a bit and see what I can come up with.

Reply:
> True, but socket authentication isn't always possible.

And in those extremely rare cases we *might* consider having the sysadmin provision the database, or write some other manual code, consider it on a case by case basis because it is so rare.

> Yes, probably. I'll have to look at the postgres module a bit and see what I can come up with.

That's great. If you need a hand ping me.

Reply:
I managed to get gitlab up and running using peer authentication and pushed a quick draft using it. The extension stuff is still TBD and I'll probably rebase this to get rid of the old stuff, but still, if you could take a look and see if it aligns with your expectations, that would be great. :)

Reply:
Only suggestion I might have is to save some typing and possibly use `optionalAttrs` on the whole thing, like [this](https://github.com/NixOS/nixpkgs/blob/c0e56afddbcf6002e87a5ab0e8e17f381e3aa9bd/nixos/modules/services/monitoring/zabbix-server.nix#L216). But yeah, exactly what I was thinking. Good stuff :+1: